system-contracts/contracts/EvmInterpreter.yul

```diff
@@ -0,0 +1,1187 @@
+object "EVMInterpreter" {
+    code { }
+    object "EVMInterpreter_deployed" {
+        code {
+            function SYSTEM_CONTRACTS_OFFSET() -> offset {
+                offset := 0x8000
+            }
+
+            function ACCOUNT_CODE_STORAGE_SYSTEM_CONTRACT() -> addr {
+                addr := 0x0000000000000000000000000000000000008002
+            }
+
+            function CODE_ADDRESS_CALL_ADDRESS() -> addr {
+                addr := 0x000000000000000000000000000000000000FFFE
+            }
+
+            function CODE_ORACLE_SYSTEM_CONTRACT() -> addr {
+                addr := 0x0000000000000000000000000000000000008012
+            }
+
+            function EVM_GAS_MANAGER_CONTRACT() -> addr {   
+                addr :=  0x0000000000000000000000000000000000008013
+            }
+
+            function DEBUG_SLOT_OFFSET() -> offset {
+                offset := mul(32, 32)
+            }
+
+            function LAST_RETURNDATA_SIZE_OFFSET() -> offset {
+                offset := add(DEBUG_SLOT_OFFSET(), mul(5, 32))
+            }
+
+            function STACK_OFFSET() -> offset {
+                offset := add(LAST_RETURNDATA_SIZE_OFFSET(), 32)
+            }
+
+            function BYTECODE_OFFSET() -> offset {
+                offset := add(STACK_OFFSET(), mul(1024, 32))
+            }
+
+            function INF_PASS_GAS() -> inf {
+                inf := 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
+            }
+
+            function MAX_POSSIBLE_BYTECODE() -> max {
+                max := 32000
+            }
+
+            function MEM_OFFSET() -> offset {
+                offset := add(BYTECODE_OFFSET(), MAX_POSSIBLE_BYTECODE())
+            }
+
+            function MEM_OFFSET_INNER() -> offset {
+                offset := add(MEM_OFFSET(), 32)
+            }
+
+            function MAX_POSSIBLE_MEM() -> max {
+                max := 0x100000 // 1MB
+            }
+
+            function MAX_MEMORY_FRAME() -> max {
+                max := add(MEM_OFFSET_INNER(), MAX_POSSIBLE_MEM())
+            }
+    
+            // It is the responsibility of the caller to ensure that ip >= BYTECODE_OFFSET + 32
+            function readIP(ip) -> opcode {
+                // TODO: Why not do this at the beginning once instead of every time?
+                let bytecodeLen := mload(BYTECODE_OFFSET())
+
+                let maxAcceptablePos := add(add(BYTECODE_OFFSET(), bytecodeLen), 31)
+                if gt(ip, maxAcceptablePos) {
+                    revert(0, 0)
+                }
+
+                opcode := and(mload(sub(ip, 31)), 0xff)
+            }
+
+            function readBytes(start, length) -> value {
```

Comment:
`length` is a constant and this function [will be specialized anyways](https://github.com/ethereum/solidity/blob/develop/libyul/optimiser/FunctionSpecializer.cpp) -- maybe you should mask here instead of using a loop?